Copter: Safe RTL

rmackay9 opened the pull request:
This replaces @squilter's earlier PR (#6682) and follow up PR (#6792) by adding support for "Safe RTL":

Improvements over the previous PRs include:

- many smallish issue resolved including changing indexes from int16_t to uint16_t
- a semaphore has been added and cleanup tasks have been moved to run in the background thread to ensure there are no conflicts between vehicle code's update (which adds points) and the background cleanup tasks
- two functions moved to AP_Math.
- prune_loops is more granular meaning it is more likely to return within 300us.

Description:

- AC_SafeRTL library does the following:
 - creates and maintains a "path" which is an array of Vector3f objects (position offsets from the EKF origin) which can lead the vehicle from its current position to the take-off location.
 - SRTL_ACCURACY parameter holds the minimum distance between points
 - SRTL_PATH_LEN holds the maximum number of points that can be recorded to the path (defaults to 150, maximum is 500)
 - the update function is called from the main vehicle class at 3.3hz and records the vehicle's position at the end of the array (if necessary)
 - the run_background_cleanup runs in the IO thread and does the following:
   - "detect_simplifications" looks for points in a straight line that can be simplified (and later removed). Points that can be removed are recorded in the _simplification_bitmask and _simplification_stack.
   - "detect_loops" looks for loops in the path which are stored in the _prunable_loops array
   - "routine_cleanup" is called when the "path" has only 10 empty slots left
   - "thorough_cleanup" is initiated when the vehicle code calls the "request_thorough_cleanup" method.  The "thorough_cleanup" ensures all possible simplification and pruning is complete so the vehicle can begin its journey home.
 - a new SafeRTL flight mode has been added which, when initiated, stops the vehicle then returns the vehicle to the take-off location by using the lib's "pop_point" method to retrieve points from the path.
- the vehicle eventually stops 2 meters above the take-off location and uses regular RTL's Land or Descend feature.

Items potential issues that we might need to address in follow-up PRs:

- we should try to optimise the segment_segment_dist and AP_Math::distance_to_segment functions.
- we should try to keep loop information even when we remove points from the path.  Currently we always throw away all info on loops even if we remove just 10 points.
- the method of marking points for removal by setting x,y,z to zero could be done more efficiently. 
- at 200 points, SafeRTL consumes about 4k of memory (2.4k for the path itself, 0.8k for each of the simplification-stack and prunable-loops arrays). Increasing the number of points consumes more memory but also more CPU for the simplification. We need to test more to find what number of points works best in practice. Same goes for the ACCURACY parameter.
- the EKF must be reporting a good position at take-off for SafeRTL to be available
- SafeRTL becomes unusable for that flight if the EKF's position estimate goes bad for >15seconds.
- SafeRTL should be made available as a failsafe option.

Some relevant links:

- Sebastian's blog post: https://discuss.ardupilot.org/t/safertl-mode-for-copter/20486
- sitl test: https://www.youtube.com/watch?v=7YsTC6vd2jE
- actual flight test: https://www.youtube.com/watch?v=Ju0tL-nmQpo

squilter commented on libraries/AP_SafeRTL/AP_SafeRTL.cpp:
```diff
@@ -0,0 +1,727 @@
+    _simplify_bitmask.setall();
+}
+
+AP_SafeRTL::~AP_SafeRTL()
```
Under what circumstances would this desctructor get called?

If it is necessary to have it, shouldn't it also `free()` the `_path`, `_prunable_loops`, `simplify_stack`, etc?

rmackay9 replied:
I'm not sure why we have a destructor.  Maybe @peterbarker knows..  I added it because I found the same thing in the AP_Airspeed library.

squilter commented:
I have an idea for a (kinda major) change to pruning. Instead of constantly comparing everything in the path to everything else, we could just compare the current position to everything else, and prune all loops instantly. It is much simpler, and solves lots of problems:

1. With a path length of 150, pruning takes like 5x the cpu power of simplification. Since pruning is O(n^2), that will only get worse when the path length is increased. But the easier simplification proposed above runs in O(n), and would probably allow for a much larger path length, maybe even making the bottleneck the memory of a pixhawk, not the cpu.

2. > we should try to optimise the segment_segment_dist and AP_Math::distance_to_segment functions.

With the proposal above, pruning could use `AP_Math::distance_to_segment` and we could remove `segment_segment_dist` entirely, which is a more expensive function.

3. > we should try to keep loop information even when we remove points from the path. Currently we always throw away all info on loops even if we remove just 10 points.

Would not be needed.

4. > the method of marking points for removal by setting x,y,z to zero could be done more efficiently.

This only exists because sometimes both algorithms need to delete points at the same time. This issue would be easy to resolve after implementing the above proposal.

It also uses a bit less memory, because there is no need for `_prunable_loops`.

The disadvantage is that certain paths would not get pruned optimally (although the current solution isn't optimal either).

muramura commented on ArduCopter/control_safe_rtl.cpp:
```diff
@@ -0,0 +1,136 @@
+        // wait for cleanup of return path
+        safe_rtl_state = SafeRTL_WaitForPathCleanup;
+        return true;
+    } else {
```
I think that the same result will be obtained without doing else statement.

rmackay9 replied:
@muramura, yes, you're right.  we can remove the extra else.  txs!

muramura commented on ArduCopter/control_safe_rtl.cpp:
```diff
@@ -0,0 +1,136 @@
+
+void Copter::safe_rtl_run()
+{
+    switch(safe_rtl_state){
```
In ardupilot, instead of "){", put whitespace in between.

rmackay9 replied:
yup, will fix!

rmackay9 commented:
@squilter, Re the pruning changes, you mean as the vehicle flies, we constantly compare the vehicle's current position to all the lines in the path?  The only issue I see is that if the vehicle is moving very quickly (i.e. 10m/s+) but we only check the positions at 3hz then we'd need to be careful that we don't miss passing over a line.

I'm not sure if it's any different than maintaining the loops even when points are removed from the path and looking into efficiency improvements in the segment-segment-dist calculations.

rmackay9 commented:
@squilter, some other questions.. (maybe we could discuss on Gitter?)

- can we remove SAFERTL_PRUNING_DELTA?  I think we could just replace the one line that uses it with _accuracy.

lvale commented:
testing



<img width="946" alt="screen shot 2017-08-31 at 03 26 43" src="https://user-images.githubusercontent.com/3130363/29903681-02e17d70-8dfd-11e7-878e-42653c535168.png">
[testmission1.plan.rename.txt](https://github.com/ArduPilot/ardupilot/files/1265600/testmission1.plan.rename.txt)

peterbarker commented on libraries/AP_SafeRTL/AP_SafeRTL.h:
```diff
@@ -0,0 +1,197 @@
+    uint16_t _simplify_stack_count; // number of elements in _simplify_stack array
+    Bitmask _simplify_bitmask = Bitmask(SAFERTL_POINTS_MAX);  // simplify algorithm clears bits for each point that can be removed
+
+    // Pruning state
```
Create a

 ```
struct {
    uint16_t i; // loop search's outer loop index
.
.
} _prune;
```

to hold the prune state.  Likewise the simplify state.

rmackay9 commented:
By the way, I've timed the two heavies calculations on a Pixhawk and the times are:

- distance_to_segment : 5.6us
- segment_segment_dist : 6.9us

I was surprised at how quick these functions are and it makes me think that it's not so important that we try to make them run faster.  Optimising so that we don't need to run them as often still makes sense - i.e. not throwing away the results when we remove points from the path is still a good thing to do.

rmackay9 commented:
I have push 4 commits that address the vast majority of items raised above by @muramura and @peterbarker.  Here's the summary:

Things fixed:
- two formatting issues brought up by muramura
- ACCURACY parameter description
- SafeRTL comments reduced to fit 80 characters
- moved a Comment down a line
- renamed definition to SAFERTL_PRUNING_LOOP_TIME_US
- comment change of "algorithm" to "algorithms"
- removed destructor (same could be done for AP_Airspeed then)
- updated GCS message to, "failed to save initial point"
- "update" comment modified from "a couple" to "several"
- added "const" ahead of "now = AP_HAL::millis()"
- added new deactivate method
- const run_background_cleanup's path_points_count
- removed unnecessary "return } else {"
- const potential_amount_to_simplify
- s/as/has in comments
- zero_points_by_simplify_bitmask starts from path[1] so as to never remove the start point
- remove_empty_points gets for loop in place of while, inverted clauses
- clarified dist_point returns value FLT_MAX in distance field
- lots of "const" added to segment_segment_dist, removed unnecessary else
- reference in Log_action.  Can't do it because want default
- ::update() comments add "3hz"

Things not fixed:
- tests in math directory
- add the descent_target.alt as a parameter to rtl_descent_start.
- better solution to 2m addition to final point (This would make it unsuitable for indoor use. Not a critical thing for this PR.)
- *s on the left in comments (it seems mixed in our code)
- didn't remove a comma in the comments but instead added another
- changing "init failed" to "allocation failed".  I don't think most users know what "allocation" would mean and "init" is shorter
- Could we use hal.scheduler->suspend_timer_procs() in the example instead?
- "s/i.e./e.g".
- consider adding "reset_simplifications" to "init".  these are already handled in "::reset" which is the only way to get the system _active.
- change return type of segment_segment_dist.

rmackay9 commented:
@lvale, any chance you've got the regular .txt format?  having problems loading it up in MAVProxy.

WickedShell commented on libraries/AP_Common/Bitmask.h:
```diff
@@ -35,6 +35,13 @@ class Bitmask {
+    // set all bits
+    void setall(void) {
+        for (uint16_t i=0; i<numwords; i++) {
+            bits[i] = 0xffffffff;
```
This is incorrect for the last word, setting all the bits could result in the unused bits of the last word (say we requested 50 bits, setting the last 14 bits is invalid). This only really matters because you wanted to calculate the hamming weight on a bitmask which this will artificially throw off.

WickedShell commented on libraries/AP_Common/Bitmask.h:
```diff
@@ -50,22 +57,42 @@ class Bitmask {
+        return sum;
+    }
+
+    // return number of bits available (may be slightly larger than number requested in constructor)
```
I'd rather this was correct. Possible uses for bitmasks include because they indirectly lookup information about something else, and iterating over every bit of a bitmask is a very natural action, and going to size makes sense. But with any indirection (IE the bitmask is used to lookup what page of memory is free) this all of a sudden could be walking off the end. If this is needed, then actually storing the count in a uint32_t makes more sense.

WickedShell commented on libraries/AP_Math/vector3.cpp:
```diff
@@ -390,6 +390,32 @@ Matrix3<T> Vector3<T>::mul_rowcol(const Vector3<T> &v2) const
+    }
+
+    // semiperimeter of triangle
+    float s = (a+b+c)/2.0f;
```
Tiny perf opt would be to make this * 0.5f

WickedShell commented on libraries/AP_SafeRTL/AP_SafeRTL.h:
```diff
@@ -0,0 +1,205 @@
+
+    // path variables
+    Vector3f* _path;    // points are stored in meters from EKF origin in NED
+    uint16_t _path_points_max;  // after the array has been allocated, we will need to know how big it is. We can't use the parameter, because a user could change the parameter in-flight
```
Worth noting that a AP_Int16 means that this is still effectively just int16_t, but if uint makes the code better then it's still worth it.

WickedShell commented on libraries/AP_SafeRTL/AP_SafeRTL.cpp:
```diff
@@ -0,0 +1,727 @@
+
+    // get semaphore
+    if (!_path_sem->take_nonblocking()) {
+        log_action(SRTL_POP_FAILED_NO_SEMAPHORE);
```
How often does this show up in logs? (SITL doesn't count given the way that SITL manages threads/background tasks at the moment)

WickedShell commented on libraries/AP_SafeRTL/AP_SafeRTL.cpp:
```diff
@@ -0,0 +1,727 @@
+// call this at 3hz (or higher) regardless of what mode the vehicle is in
+void AP_SafeRTL::update(bool position_ok, bool save_position)
+{
+    if (!_active || !position_ok || !save_position) {
```
Checking `position_ok` here seems like a waste, since the general case is that it will be, we then and on to it, and check again in the next function call, why not only check it then?

WickedShell commented on libraries/AP_SafeRTL/AP_SafeRTL.cpp:
```diff
@@ -0,0 +1,727 @@
+    }
+
+    // check for timeout due to bad position
+    if (AP_HAL::millis() - _last_good_position_ms > SAFERTL_TIMEOUT) {
```
This should be the else case on the previous `if (position_ok) {` as entering that if will mean this one is never hit.

WickedShell commented on libraries/AP_SafeRTL/AP_SafeRTL.cpp:
```diff
@@ -0,0 +1,727 @@
+    }
+
+    // check for timeout due to failure to save points to path (most likely due to buffer filling up)
+    if (AP_HAL::millis() - _last_position_save_ms > SAFERTL_TIMEOUT) {
```
This might as well be done in the else case of add_point as thats the only time we could hit this

WickedShell commented on libraries/AP_SafeRTL/AP_SafeRTL.cpp:
```diff
@@ -0,0 +1,727 @@
+    _path_points_max = _points_max;
+
+    // create semaphore
+    _path_sem = hal.util->new_semaphore();
```
Technically can be null if we perfectly used up memory, might as well do this before checking all the other nullptr's

WickedShell commented on libraries/AP_SafeRTL/AP_SafeRTL.cpp:
```diff
@@ -0,0 +1,727 @@
+    _thorough_clean_complete_ms = 0;
+
+    // check if path array is nearly full, if yes we should do a routine cleanup (i.e. remove 10 points)
+    bool path_nearly_full = path_points_count >= MAX(_path_points_max - SAFERTL_CLEANUP_START_MARGIN, 0);
```
path_points_count would need to be refetched after the thorough cleanup may have been run as it may have changed.

rmackay9 commented:
@squilter, for optimising CPU usage I think we should consider doing this:

- simplify after every X points (50?) have been added, immediately remove the simplify-able points from the path and never try to simplify this group of points again.  We could avoid looking at them again by recording the index of the last point simplified in a new variable (i.e. simplify.previous_top?).  On the next iteration we would start from the index held in this new variable (or perhaps this index-1).
- prune should be restructured so at least the outer loop checks from high indices to low.
- prune right after simplification has run and only on the simplified points.  This means we will have the min number of points and we know that these loops will never be affected by simplification which will mean when we prune loops, we can more easily adjusted the indexes of existing prune loops and avoid repeated searches.

squilter commented:
@rmackay9 

> can we remove SAFERTL_PRUNING_DELTA? I think we could just replace the one line that uses it with _accuracy.

No, it must be slightly less than _accuracy. Imagine a bunch of points in a straight line, all 2m from each other. The loop detector would detect a loop between 1&3, 2&4, 3&5, etc.

> does distance-to-segment properly respect ends of lines or is it perpendicular distance to line?

It's the perpendicular distance to the line, not the line segment. That's why the method was originally named `_point_line_dist`. It should probably be renamed.

> prune should be restructured so at least the outer loop checks from high indices to low.

There are pros and cons for each way. I never decided which one was better.

> prune right after simplification has run and only on the simplified points.

I think this would work. Let me think about it some more. The biggest problem that I see is that if there have been 49 points added since the last simplification when the user switches to SafeRTL mode, the user is going to have to wait while the algorithms simplify the path and then prune loops with the new points, which might mean that the copter hovers in place for a few seconds before returning.

squilter commented:
re: your idea

For the simplification algorithm to produce the correct answer, it must take the whole path as an input. It is not correct to split the path into smaller segments, simplify those, and concatenate them. The result of doing that probably wouldn't be horrible, it just wouldn't be as good. [Here](https://en.wikipedia.org/wiki/Ramer%E2%80%93Douglas%E2%80%93Peucker_algorithm#/media/File:Douglas-Peucker_animated.gif) is an animation of how the algorithm works to help understand the limitations.

squilter commented:
Fixes to Bitmask.h recommended by @WickedShell are at squilter/squilter-safe-rtl8

rmackay9 commented:
@squilter, I've pulled in the bitmask fix, thanks!

I'm not sure how important it is but I've seen this difference discussed on some web pages re distance-to-line and distance-to-segment.  I suspect in practice it doesn't make much difference but my understanding is that distance-to-segment is what we should use.. what do you think?

rmackay9 commented:
@WickedShell, thanks again for the review.  It's really helpful.  I've added a few commits from @squilter and I that addresses the issues I think.

- bitmask fixed
- replace divide with multiply in AP_Math
- check semaphore allocation
- fixup update method (remove redundant position_ok, re-order timeout checks)

Some issues not addressed:

- how often does the "SRTL_POP_FAILED_NO_SEMAPHORE" failure show up in real flight logs?  I'm not sure actually so we will test.
- path_points_count would need to be refetched after the thorough cleanup and before routine-cleanup.  Actually I don't think it's necessary because there's a return after thorough-cleanup.

gmorph commented on ArduCopter/ArduCopter.cpp:
```diff
@@ -64,6 +64,7 @@
  *  Roberto Navoni      :Library testing, Porting to VRBrain
  *  Sandro Benigno      :Camera support, MinimOSD
  *  Sandro Tognana      :PosHold flight mode
+ *  Sebastian Quilter   :SafeRTL
```
Update to SmartRTL now I assume?

gmorph commented on ArduCopter/control_safe_rtl.cpp:
```diff
@@ -0,0 +1,136 @@
+        Vector3f next_point;
+        if (g2.safe_rtl.pop_point(next_point)) {
+            if (g2.safe_rtl.get_num_points() == 0) {
+                // this is the very last point, add 2m to the target alt and move to pre-land state
```
Why are we adding 2m to the alt?  I might be perfectly valid but I've no idea why.  Can we improve the comment to clarify why this is done?

gmorph commented on ArduCopter/control_safe_rtl.cpp:
```diff
@@ -0,0 +1,136 @@
+            safe_rtl_state = SafeRTL_Land;
+        } else {
+            rtl_path.descent_target.alt = g.rtl_alt_final;
+            rtl_descent_start();
```
The 2m value looks like a magic number in copter.  I'm guessing the default safety margin.

gmorph commented on ArduCopter/system.cpp:
```diff
@@ -227,6 +227,9 @@ void Copter::init_ardupilot()
 
+    // initialize SafeRTL cleanup methods
+    g2.safe_rtl.init();
+
```
Is this ok to initialise before the dataflash?

gmorph commented on libraries/AP_Common/Bitmask.h:
```diff
@@ -20,21 +20,37 @@
 
     // set given bitnumber
     void set(uint16_t bit) {
+        if (bit >= numbits) {
```
Might be worth a comment stating ignoring setting an invalid bit number.  Ok - so I'm being fussy.

gmorph commented on libraries/AP_Common/Bitmask.h:
```diff
@@ -44,29 +60,48 @@ class Bitmask {
-        for (uint16_t i=0; i<numwords; i++) {
-            bits[i] = 0;
-        }
+        memset(bits, 0, numwords);
```
I think this is wrong.  bits is an int32_t but we are only setting the first numwords bytes of it.  Shouldn't this be 
memset(bits, 0, sizeof(bits))

tridge replied:
it is a bug, but sizeof(bits) would be worse, as that is a pointer.
The original code was correct, or you could do numwords*sizeof(bits[0])

gmorph replied:
hahahaha - of course.  Thanks Tridge.

rmackay9 replied:
@squilter, can you check this one?  If it's just a matter of reverting it I can do that.. or you can just make another commit.  whatever works.

gmorph commented on libraries/AP_Common/Bitmask.h:
```diff
@@ -44,29 +60,48 @@ class Bitmask {
+    uint16_t count() const {
+        uint16_t sum = 0;
+        for (uint16_t i=0; i<numwords; i++) {
+            if (sizeof(bits[i]) <= sizeof(int)) {
```
shouldn't this be int16_t?

squilter replied:
No, I think this is correct. `__builtin_popcount()` takes an `unsigned int`, `__builtin_popcountl()` an `unsigned long` and `__builtin_popcountll()` an `unsigned long long`.

gmorph commented on libraries/AP_Common/Bitmask.h:
```diff
@@ -44,29 +60,48 @@ class Bitmask {
+        for (uint16_t i=0; i<numwords; i++) {
+            if (sizeof(bits[i]) <= sizeof(int)) {
+                sum += __builtin_popcount(bits[i]);
+            } else if (sizeof(bits[i]) <= sizeof(long)) {
```
and this int32_t?

gmorph commented on libraries/AP_SafeRTL/AP_SafeRTL.cpp:
```diff
@@ -0,0 +1,728 @@
+
+    _path_points_max = _points_max;
+
+    // when running the example sketch, we want the cleanup tasks to run when we tell them to, not in the background (so that they can be timed.)
```
wots a sketch?

squilter replied:
[Library example sketch](http://ardupilot.org/dev/docs/learning-ardupilot-the-example-sketches.html)

gmorph commented:
Excellent PR.  Extremely clean code - a pleasure to read.  Nicely done.

rmackay9 commented:
@gmorph, thanks for the review!

- I think we will leave it called "SafeRTL" for now.  we can rename it in master if that's what everyone wants.  It's @squilter's baby so we should probably attach a heavy weighting to what he wants to call it.
- re the 2m added to the final point when returning (i.e. the take-off location), @peterbarker brought this up as well but we don't have a better idea of what to do just yet.  We can't have the copter flying straight towards the take-off location because it would likely scrape the ground and flip-over.  So at the moment we aim for 2m above the take-off location.
- happy to change those comments
- I've leave it to @squilter to address the other BitMask issues.

@squilter, I think we should address these final items from @gmorph and @tridge and then I'll likely push this into master (hopefully tomorrow).  You may have noticed that I'm attempting a CPU optimisation in my repo's https://github.com/rmackay9/rmackay9-ardupilot/tree/squilter-safe-rtl9 branch.  It's still a WIP though and I think it can be a follow-up PR if it works and you and others are happy with it.

squilter commented:
Okay, I made that one bitmask fix, feel free to squash that commit in somewhere.

Regarding dist_line_point vs dist_segment_point, you are right, we should be using segment. Here's an idea for how to do it, but it seems pretty expensive:

(Point is called A, segment endpoints are B and C)

```
if the angle ABC is > 90°:
    return HYPOT(A,B)
else if the angle BCA is > 90°:
    return HYPOT(A,C)
else:
    return point_line_dist
```

Where the angle calculation is accomplished with

![screenshot from 2017-09-01 18-09-29](https://user-images.githubusercontent.com/2956460/29989429-ce8af3f4-8f40-11e7-9952-446daa45b45c.png)
(copied from [here](https://math.stackexchange.com/questions/361412/finding-the-angle-between-three-points))

I think this is one of those problems that has a hundred different answers. Can anybody think of a faster one?

squilter commented:
SafeRTL, SmartRTL, smaRTL, etc. I have no strong preference.

lvale commented:
Replicate a flight from today with a mission and tested several activations of safeRTL/SmartRTL with some unexpected results.

Frightening on the field was the descent rate to home, that can also be seen on this SITL

[Archive.zip](https://github.com/ArduPilot/ardupilot/files/1271741/Archive.zip)

rmackay9 commented:
@lvale, thanks for testing although the -rtl9 branch is totally a work-in-progress.  -rtl8 (the one in this PR is safe) but -rtl9 is absolutely not working!

rmackay9 commented:
Above changes are just squashing in the Bitmask fix from @squilter.

@squilter, I think we have an issue with overlapping loops.  I don't think we end up with a valid path if we try to remove points from two loops that overlap.  Instead what I think we need to do is only remove the points from the longer loop (which might not be the one with more points).  I'm going to see if I can tackle this I think.

I'm not certain but I think this is the cause of the issue that @lvale has seen (and I've reproduced).  Basically what we're seeing is that sometimes the vehicle is cutting across the circles in @lvale's mission - flying in places where it's never flown before.

rmackay9 commented:
@squilter, this problem gets harder the longer I look at it.  I think now that a fundamental limitation of the current method is the vehicle must travel backwards along all segment (i.e. in the reverse direction that it flew in before SafeRTL was initiated).  There are cases where this is not the shortest path when we have overlapping loops (see pic on the right below).  Ideally the vehicle should return from end -> 51/22 -> 25/13 -> 15/10 -> start but we will never allow this because 22->25 is in the wrong direction.
![loop-issue-pic](https://user-images.githubusercontent.com/1498098/29992265-741b8d1a-8fd2-11e7-914d-5744d1add728.png)

rmackay9 commented:
I've made the bulk of the changes to reduce the CPU load and resolve the issue Luis has reported: https://github.com/rmackay9/rmackay9-ardupilot/commits/squilter-safe-rtl9

I need to test it further however because there seems to be a bug with the pruning which can be seen in this video: https://www.youtube.com/watch?v=z6ZnZbStEsE.  Still, it performs in a safe manner than -rtl8 did.

squilter commented:
@rmackay9 The issue of overlapping loops gets complicated when there are "chains" of overlapping loops, because then (as you said) it is sometimes fastest to fly a certain segment backwards.

I had an imperfect, but decent solution. I think I stripped it out while debugging one day and then forgot to bring it back for the initial PR.

- Allow `detect_loops()` to detect once-overlapping loops. (e.g. detecting loops from 1->5, 3->8,6->10 is ok)
- Do Not allow `detect_loops()` to detect a loop-within-a-loop (e.g. detecting loops from 1->6 and 2->5 is **not** ok)
- Do Not allow `detect_loops()` to detect twice-overlapping loops. (e.g. detecting loops from 1->6 and 2->8 and 4->10 is **not** ok)

Immediately before pruning all loops, run the method below. What it does is it finds all loop-chains and removes either all the even-numbered loops from the chain or all the odd-numbered loops (leaving no overlapping loops). Then it is safe to prune all remaining loops from `_prunable_loops`.  This is the best result that can be achieved without flying backwards along a segment and without allowing the detection of twice-overlapping (or more) loops. Obviously these two restrictions can make it produce a sub-optimal result, but it was the best I could think of.

```
/**
*   _detect_loops never returns loops-within-loops, but sometimes it does return overlapping loops.
*   This method looks through the loops, and wherever there is a chain of overlapping loops, it
*   either removes all even-numbered-loops or all odd-numbered loops, whichever produces the shorter path.
*/
void SafeRTL_Path::_remove_unacceptable_overlapping_loops()
{
    if (_prunable_loops.size() == 0) {
        return;
    }

    // iterate through all loops. dont increment i here, that happens inside
    uint32_t i;
    for (i = 0; i < _prunable_loops.size() - 1;) {
        uint32_t chain_len = 0;
        uint32_t j;
        for (j = i+1; j < _prunable_loops.size() - 1; j++) {
            // if two loops overlap,
            if (_prunable_loops[j-1].end_index > _prunable_loops[j].start_index) {
                chain_len++;
            } else {
                break;
            }
        }
        // if no loop was detected
        if (chain_len == 0) {
            i = j;
            continue;
        }
        // if we got here, we have identified a chain of overlapping loops,
        // starting with loop at index i, ending at i+chain_len
        // now, decide if it would be better to delete even or odd loops
        float even_dist_removed = 0;
        float odd_dist_removed = 0;
        // for each prunable loop in the overlapping chain
        uint32_t k;
        for (k = i; k <= i + chain_len; k++) {
            float loop_len = 0;
            // for each point in said prunable loop
            for (uint32_t l = _prunable_loops[k].start_index; l < _prunable_loops[k].end_index; l++) {
                // add the real-world distance between two sequential points in the loop
                loop_len += HYPOT(path[l-1], path[l]);
            }
            // now add the real-world length of that loop to the correct counter
            if (k%2==0) { // evens
                even_dist_removed += loop_len;
            } else {   //odds
                odd_dist_removed += loop_len;
            }
        }
        // if we want to delete evens, but the loops starts at odds, or if we want to delete odds but the loop starts at evens, offset deletion by one.
        bool delete_offset = (even_dist_removed < odd_dist_removed) == (k % 2);
        for (uint32_t l = i + delete_offset; l <= i + chain_len; l++) {
            _prunable_loops.erase(_prunable_loops.begin() + l);
            chain_len--;
        }
        i = j; // next iteration, only start looking for the next chain at an index greater than the end of this chain
    }
}
```

rmackay9 commented:
Closing in favour of the replacement PR (https://github.com/ArduPilot/ardupilot/pull/6891)
